Spec.md

```diff
@@ -0,0 +1,227 @@
+- Provide type information
+- Methods should be considered
+- Online detection (possibly as separate topics)
+- Allow empty objects
+- Events/alarms must be mappable
+- Coordinate the completeness of the run
+- Encoding should be UA JSON
+- Dimensioning for approximately 350 subscribers (machines + dashboard users)
+- Custom DataTypes must be included
+- The address space includes at least OPC UA Machinery and possibly other Companion Specifications based on it.
+- The entry point of the mapping is the machine object in the Machines folder of Machinery.
+
+## Design Decisions
+
+The concept of OPC UA Part 14 should be followed as closely as possible.
+
+## Mapping
+
+### Mapping of Objects and Variables
+
+- An object is mapped to a DataSet.
+- A variable is mapped as a field to the DataSet. If the variable contains children (other variables), a DataSet with the children is created.
+
+#### Example
+
+![image](https://github.com/umati/UA-CloudPublisher/assets/12342602/91eae1d7-9d6f-4246-abb2-0ec5ea9b59bc)
+
+The image shows a part of an address space. Each object is mapped to a DataSet:
+
+- Production
+- ActiveProgram
+- State
+
+The properties Name, NumberInList, Id, and Number are fields of the DataSets. The value of CurrentState is mapped as a field of State and has a DataSet with its properties:
+
+- Production
+- ActiveProgram
+  - Name
+  - NumberInList
+  - CurrentState
+- State
+- CurrentState
+  - Id
+  - Number
+
+### Explanation
+
+A field can also contain properties but no other variables, even though the address space can contain more hierarchies. Properties of objects also cannot be mapped to the properties of a field. Therefore, a uniform mapping is used for objects and variables so that access is always consistent.
+
+## Mapping of References
+
+The name of a DataSet is created by the BrowsePath of the object. The NamespaceUri is used instead of the NamespaceIndex to make the path server-independent. Other BrowsePaths may also be possible. Alternative BrowsePaths can be mapped to the description (see description field of the MetaData). References can also be mapped into the description.
+
+## Topic Structure
+
+The generic topic structure for OPC UA is:
+
+`<Prefix>/<Encoding>/<MqttMessageType>/<PublisherId>/[<WriterGroup>[/<DataSetWriter>]]`
+
+For the umati Dashboard, this topic structure is restricted as follows:
+
+- `<Prefix> = umati.v3`
+- `<Encoding> = json`
+- `<MqttMessageType> = as defined in OPC UA PubSub`
+- `<PublisherId> = Name of the Client`
+- `<WriterGroup> = Expanded NodeId of the object representing the machine`
+- `<DataSetWriter> = BrowsePath`
+
+The BrowsePath in the DataSetWriter is built with a "." between each BrowseName. The BrowseNames use only the namespace index to avoid collisions.
+Special character as `'/''.''<'| '>' | ':' | '#' | '!' | '&' ` in the `name` field of the BrowseName  need to encode by [URL-Encoding](https://de.wikipedia.org/wiki/URL-Encoding) using an underscore instead of a '%'.

```

Comment:
@ccvca Do you think this solves the problem with the encoding of the BrowsePath?

Reply:
Should work, but `_` must be in the list, too.  And the escaped characters must be either specified directly, or a Client must not guess the topics without reading the metadata, as some Servers may include more characters than others. (e.g. the current client escapes all characters except [a-z0-9].

Reply:
escaping all characters except [a-z0-9] seems a good idea.